allow pbjs to be used as external module

olafbuitelaar opened the pull request:
added some changed so pbjs could be used as module as general, and added some enhancements. some might conflict with design principles, but not sure about that.

mkendall07 commented:
@olafbuitelaar 
Can you explain the goals of this PR? Also I see a lot of unrelated changes. Please limit a PR to a single change.

olafbuitelaar commented:
@mkendall07
Sure, this PR fixes in my opinion 5 issues;

1)	pbjs clears all bidding data upon a requestBids call.  This causes all calls to renderAd to become invalid for the removed bids if a second call to requestBids was done. Ideally the auction process and render process would be separate. In theory what could happen is that the http adserver response with the call the renderAd is still in transit, while the client already performed a second call to requestBids. Initially it might not seem so prominent, but with the web evolving to be more async this might surface. Especially with the rising of the infinite scrolling pages and SPA’s. In addition having them separate opens the possibility to apply client side logic on the received bids. 

2)	The adUnitBidsBack callback wasn’t called reliably when an adapter timed out. As the request count didn’t match the received count when a timeout occurred.

3)	The adUnitBidsBack callback does now receive the adUnitCode, so the callback can determine the adunit for which the bids are back. 

4)	The pbjs rubicon adapter can now be ran in its own iframe

5)	Giving a simple implementation so removeCallback can at least can remove the added callbacks.

A sample of it running can be found here; http://cdn.adsolutions.nl/tagman/samples/prebid/prebid.html 

Given the PR is a bit rough, but I wasn’t sure where the pbjs community would stand.
I’m happy to polish it up a bit, make it follow project standards.

mkendall07 commented:
1) Should be resolved by https://github.com/prebid/Prebid.js/pull/477
2 & 3) It would be great to submit as a separate PR together. 
4) Why does the Rubicon adapter need to run in an iframe?
5) Please submit as separate PR.

In general, we limit a PR to a single change and it should be accompanied by unit tests in spec files. Thank you.

olafbuitelaar commented:
@mkendall07

1)	Yes it appears we both worked out a solution for the same problem. However I think both solutions aren’t ideal as it would be impossible to know if a bit is ready to be removed, only when a call to renderAd is done you know you could remove the history, in any other case you don’t know if the renderAd call is still in transit or failed, or simply if the adserver just didn’t pick the prebid campaign and delivered something else instead. Personally I would favor keeping the request/response queue as small/clean as possible, instead of maintaining the state there, at the cost of the double array. However I can see the cleanness #477 in it as well. For both I would say the bids needs to have some kind of lifetime after completion, to be able to clean them.
2)	& 3) it appears my commit isn’t fully atomic, I’ll see if it’s easy to untangle them. Or is it easy to cherry pick them?
4)	Maybe I should rephrase, I meant running pbjs in a iframe, if I have this setup rubicon places the rubicontag global in the iframe pbjs runs in. if I would later pick a rubicon as selected winner and render the ad in its own iframe it’s impossible for the generated code to resolve the rubicontag instance. Initially I just redistributed the variable, but later I’ve made an option for the adapter to pre initialize itself for delivery so it has a chance to manage it’s own globals, see commit; #7459f38
5)	Same as 2) & 3)

protonate commented on src/adaptermanager.js:
```diff
@@ -132,6 +132,17 @@ exports.enableAnalytics = function (config) {
 };
 
+//if an adapter needs to prepare anything before the ad can be delivered, give it a chance to do so.
+exports.prepareRendering = function (doc, bidderCode) {
```
This may prove useful to other adapters but until we have a need for it the setup Rubicon needs to work with iframes and globals can be kept in the adapter itself. Also `adaptermanager` is used when preparing to request bids and `bidmanager` is used to handle the bid responses, so preparing to render seems to belong in the later module.

protonate commented on src/adapters/appnexus.js:
```diff
@@ -186,6 +185,7 @@ AppNexusAdapter = function AppNexusAdapter() {
         bid.width = jptResponseObj.result.width;
         bid.height = jptResponseObj.result.height;
         bid.dealId = jptResponseObj.result.deal_id;
+        bid.bidId = id;//set bidId, so later on the correct response can be matched with the request, if AN is in the race multiple times for the same adUnit
```
As mentioned elsewhere this is achieved in #477 by passing the origin bid request to `bidfactory.createBid`.

protonate commented on src/adapters/appnexus.js:
```diff
@@ -198,6 +198,7 @@ AppNexusAdapter = function AppNexusAdapter() {
         //indicate that there is no bid for this placement
         bid = bidfactory.createBid(2);
         bid.bidderCode = bidCode;
+        bid.bidId = id;//set bidId, so later on the correct response can be matched with the request, if AN is in the race multiple times for the same adUnit
```
Same as comment for L188

protonate commented:
There are a couple places where changes in this PR are recently made in other PRs, and actually this branch now has conflicts do to these overlapping changes. Please update your branch to resolve.

olafbuitelaar commented:
@mkendall07 

2 & 3) I’ve create separate commits for it (#e05fc09 and #b915dfb), but it seems I’m unable to create pull requests for it.

5) same, it can be found at commit #3810a9b

@protonate
I’ve updated the AppNexus adapter to use the new createBid interface. I didn’t want/dare to mangle with the adId, so chose the bidId property instead, as adId already was used. Also by the time I implemented it, that code wasn’t available yet in the main branch. 

I understand your concerns about the place where prepareRendering would live. Obviously I’m new to the code base. For me it seemed to most logical place, as the bidmanger doesn’t seem to have direct access to the adapters. Sure it could access bidderRegistry (but I didn’t see that pattern anywhere in the code), Than I would rather feel the adaptermanager should have a method to resolve an given adapter from a bid. Personally I interpreted the adaptermanager as a gateway to the adapters, not so much limited to only the callBids call. But appears that assumption was wrong. However if it would live in the bidmanager the call would be; prebid=>renderAd -> bidmanager=>prepareRendering (adaptormanager=>resolveAdapterFromBid) -> adapter=>prepareRendering, which seems a long way.

protonate commented:
On Friday it was necessary to remove a commit from Prebid.js, and a force push of master branch has rewritten history. As a committer with an open PR this event impacts you, and we regret the inconvenience. We will need to close your PR and ask that you reapply your changes to a current fork of master, then open a new PR. This will align your history with the main project and your PR will show only your changes. 

Committers are referred to the Apache License (https://github.com/prebid/Prebid.js/blob/master/LICENSE) for details on submitting code to the Prebid.js project.
